### PR TITLE
#82: ClassCastException after Plugin update on DC

### PR DESCRIPTION
**Pull request by Tiliavir**
Besides other posts regarding this issue in the "Atlassian Developer Community" [this one](https://community.developer.atlassian.com/t/cache-in-data-center-throws-classcastexception-when-casting-to-the-same-class/3232/4) summarizes it quite well:

> [...] is caused by the fact during cache replication in Data Center RMI uses webapp class loader to deserialize cache invalidation message and not OSGi class loader associated with your plugin’s OSGi bundle.
> The easiest way to work around this problem is to use keys that can be serialized/deserialized by webapp class loader (e.g. java.lang.String) [...]

and further

> The Atlassian guide on using Data Centre Compatible caches [here 19](https://developer.atlassian.com/server/confluence/how-do-i-cache-data-in-a-plugin/) says the cache can contain keys and values that implement Serializable however from what I have seen serializing pojos does not work as there are different classloaders in play which results in ClassCastException.

Following that, I've added this PR that should address this issue in a self healing way. Meaning: The issue might occur again after an environment gets this patch (and requires a manual plugin cache reset / reboot once again), but should be fixed for all following updates.

### CODE REVIEW

File: src/main/java/com/baloise/confluence/digitalsignature/DigitalSignatureMacro.java [hunk 1]
```diff
@@ -42,338 +42,331 @@
...
+      // required for downward compatibility - update for next time.
+      Signature signature = (Signature) value;
+      this.bandanaManager.setValue(context, key, signature.serialize());
+      return signature;
     }
 
-    private void save(Signature signature) {
-        if (signature.hasMissingSignatures())
-            bandanaManager.setValue(GLOBAL_CONTEXT, signature.getKey(), signature);
+    if (value instanceof String) {
```
Comment (arburk) [inline src/main/java/com/baloise/confluence/digitalsignature/DigitalSignatureMacro.java:324]: 👍

**Review by arburk — APPROVED** [review]

**Comment by MarkusTiede** [comment]
@Tiliavir asks to review this commit : https://github.com/baloise/digital-signature/pull/88/commits/b86a2cf4e83a8c5a76a045665bbf86d86b5aa3d1

**Comment by Tiliavir** [comment]
Tests to be performed before release:
[ ] Upgrade from previous version - ensure signatures are still there
[ ] Retest NPE from #77, #80

After Release
[ ] Update #77, #80, #82
[ ] Update version in Atlassian Market Place
[ ] Inform users about new version

**Comment by Tiliavir** [comment]
debuggen with:
  docker run -e JVM_SUPPORT_RECOMMENDED_ARGS="-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005" -v confluenceVolume:/var/atlassian/application-data/confluence --name=confluence -d -p 5005:5005 -p 8090:8090 -p 8091:8091 atlassian/confluence-server:latest

- test deploy undeploy deploy
  if ok - test update
- test deploy

**Pull request merged**